flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/statistics/basicstatistic/MultivariateGaussian.java

```diff
@@ -0,0 +1,148 @@
+	private static final double EPSILON;
+
+	static {
+		double eps = 1.0;
+		while ((1.0 + (eps / 2.0)) != 1.0) {
+			eps /= 2.0;
+		}
+		EPSILON = eps;
+	}
+
+	private final DenseVector mean;
+	private final DenseMatrix cov;
+
+	private DenseMatrix rootSigmaInv;
+	private double u;
+
+	// data buffers for computing pdf
+	private DenseVector delta;
+	private DenseVector v;
+
+	/**
+	 * The constructor.
+	 *
+	 * @param mean The mean vector of the distribution.
+	 * @param cov  The covariance matrix of the distribution.
+	 */
+	public MultivariateGaussian(DenseVector mean, DenseMatrix cov) {
+		this.mean = mean;
+		this.cov = cov;
+		this.delta = DenseVector.zeros(mean.size());
+		this.v = DenseVector.zeros(mean.size());
+		calculateCovarianceConstants();
+	}
+
+	/**
+	 * Returns density of this multivariate Gaussian at given point, x.
+	 */
+	public double pdf(Vector x) {
+		return Math.exp(logpdf(x));
+	}
+
+	/**
+	 * Returns the log-density of this multivariate Gaussian at given point, x.
+	 */
+	public double logpdf(Vector x) {
+		int n = mean.size();
+		System.arraycopy(mean.getData(), 0, delta.getData(), 0, n);
+		BLAS.scal(-1.0, delta);
+		if (x instanceof DenseVector) {
+			BLAS.axpy(1., (DenseVector) x, delta);
+		} else if (x instanceof SparseVector) {
+			BLAS.axpy(1., (SparseVector) x, delta);
+		}
+		BLAS.gemv(1.0, rootSigmaInv, false, delta, 0., v);
+		return u - 0.5 * BLAS.dot(v, v);
+	}
+
+	/**
+	 * Compute distribution dependent constants.
+	 *
+	 * <p>The probability density function is calculated as:
+	 * pdf(x) = (2*pi)^(-k/2)^ * det(sigma)^(-1/2)^ * exp((-1/2) * (x-mu).t * inv(sigma) * (x-mu))
+	 *
+	 * <p>Here we compute the following distribution dependent constants that can be reused in each pdf computation:
+	 * A) u = log((2*pi)^(-k/2)^ * det(sigma)^(-1/2)^)
+	 * B) rootSigmaInv = sqrt(inv(sigma)) = D^(-1/2)^ * U.t
+	 *
+	 * <p><ul>
+	 * <li> sigma = U * D * U.t
+	 * <li> inv(sigma) = U * inv(D) * U.t = (D^{-1/2}^ * U.t).t * (D^{-1/2}^ * U.t)
+	 * <li> sqrt(inv(sigma)) = D^(-1/2)^ * U.t
+	 * </ul></p>
+	 */
+	private void calculateCovarianceConstants() {
+		int n = this.mean.size();
+		int lwork = 3 * n - 1;
+		double[] matA = new double[n * n];
+		double[] work = new double[lwork];
+		double[] evs = new double[n];
+		intW info = new intW(0);
+
+		for (int i = 0; i < n; i++) {
+			System.arraycopy(cov.getData(), i * n, matA, i * n, i + 1);
+		}
+		LAPACK_INST.dsyev("V", "U", n, matA, n, evs, work, lwork, info);
+
+		double maxEv = -Double.MAX_VALUE;
+		for (int i = 0; i < n; i++) {
+			if (evs[i] > maxEv) {
+				maxEv = evs[i];
+			}
+		}
+		double tol = EPSILON * n * maxEv;
+
+		// log(pseudo-determinant) is sum of the logs of all non-zero singular values
+		double logPseudoDetSigma = 0.;
+		for (double ev : evs) {
+			if (ev > tol) {
+				logPseudoDetSigma += Math.log(ev);
+			}
+		}
+
+		this.rootSigmaInv = new DenseMatrix(n, n);
+		for (int i = 0; i < n; i++) {
```

Comment:
nit: this seems to be able to abstract into a subroutine in `BLAS` ? (I am not sure whether this is useful but good to check?

Reply:
I think it could be abstracted as a full matrix multiplies a diagonal matrix. Perhaps we can add to BLAS in another PR.